File: web-face/src/app/core/services/tag.service.ts
import { inject, Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

import { environment } from '../../../environments/environment';
import { buildHttpParams } from './http-params.util';

/** Free-text tag autocomplete, aggregated across Location and Device. */
@Injectable({ providedIn: 'root' })
export class TagService {
  private readonly http = inject(HttpClient);

  suggest(prefix: string, limit = 10): Observable<string[]> {
    return this.http.get<string[]>(`${environment.apiUrl}/tags`, {
      params: buildHttpParams({ prefix, limit }),
    });
  }
}

File: core/src/main/java/com/csl/lasform/service/TagService.java
package com.csl.lasform.service;

import java.util.List;

/** Free-text tag autocomplete, aggregated across every entity that carries a {@code tags} field. */
public interface TagService {

    /** Existing tags (across Location and Device) starting with {@code prefix}, case-insensitively. */
    List<String> suggest(String prefix, int limit);
}
